pacai/agents/search/base.py

```diff
@@ -16,18 +19,29 @@ class SearchAgent(BaseAgent):
     """
 
     def __init__(self, index,
-            fn = 'pacai.student.search.depthFirstSearch',
-            prob = 'pacai.core.search.position.PositionSearchProblem',
-            heuristic = 'pacai.core.search.heuristic.null',
+            fn: Union[str, Callable[[SearchProblem], any]]
+                = 'pacai.student.search.depthFirstSearch',
+            prob: Union[str, Callable[[AbstractGameState], SearchProblem]]
+                  = 'pacai.core.search.position.PositionSearchProblem',
+            heuristic: str = 'pacai.core.search.heuristic.null',
             **kwargs):
         super().__init__(index, **kwargs)
 
-        # Get the search problem type from the name.
-        self.searchType = reflection.qualifiedImport(prob)
+        if isinstance(prob, str):
+            # Get the search problem type from the name.
+            self.searchType = reflection.qualifiedImport(prob)
+        else:
+            # Use provided problem or problem factory
+            self.searchType = prob
         logging.info('[SearchAgent] using problem type %s.' % (prob))
 
-        # Get the search function from the name and heuristic.
-        self.searchFunction = self._fetchSearchFunction(fn, heuristic)
+        if isinstance(fn, str) and isinstance(heuristic, str):
+            # Get the search function from the name and heuristic.
+            self.searchFunction = self._fetchSearchFunction(fn, heuristic)
+        else:
+            # Use provided search function and ignore heuristic
+            logging.info('[SearchAgent] using function %s.' % (fn))
```

Comment:
Move this log outside of the if/else, and use `self.searchFunction` (mirror the problem type log statement).

Reply:
This will log twice because of `_fetchSearchFunction`

Reply:
You can feel free to edit `_fetchSearchFunction` as well so that the logs are clear.